### PR TITLE
Prevent signal handler blocking

### PR DESCRIPTION
**Pull request by xmo-odoo**
Since Python only allows signal handlers on the main thread and has a big lock, some events (e.g. IO, accept(2), ...) can defer the execution of signal handlers. If the event in question gets stuck, it makes e.g. dumpstack (`kill -3`) unresponsive which is inconvenient to diagnose stuck processes.

By moving the potentially blocking work to a sub-thread and idling the main thread, the event handlers should always run.

Task 1913129

### CODE REVIEW

**Comment by xmo-odoo** [comment]
@antonylesuisse @odony @beledouxdenis as part of the PR I had to move some of the bits from `process_limit` (part of the worker process's runloop) to `start`, but I'm not quite clear why they were part of the runloop in the first place as they're setting a signal handler (sigxcpu) and setting a completely static resource limit (memory).

Also `process_limit` is the first task of the runloop, therefore we're still going to wait for and run a work item even if already above the soft limits, why not run `process_limit` after `process_work`?

**Comment by odony** [comment]
> Also `process_limit` is the first task of the runloop, therefore we're still going to wait for and run a work item even if already above the soft limits, why not run `process_limit` after `process_work`?

@xmo-odoo I guess it's because `process_limit()` also sets the limits, so it had to be done before `process_work()` (e.g for RLIMIT_CPU). We could indeed split it into set_limits() and check_limits() if we want to get an earlier suicide.

File: odoo/service/server.py [hunk 1]
```diff
@@ -424,7 +424,7 @@ def run(self, preload=None, stop=False):
         """
         self.start(stop=stop)
 
-        rc = preload_registries(preload)
```
Comment (antonylesuisse) [inline odoo/service/server.py]: Not needed.

File: odoo/service/server.py [hunk 2]
```diff
@@ -742,7 +742,7 @@ def stop(self, graceful=True):
     def run(self, preload, stop):
         self.start()
 
-        rc = preload_registries(preload)
```
Comment (antonylesuisse) [inline odoo/service/server.py]: not needed

File: odoo/service/server.py [hunk 3]
```diff
@@ -856,17 +849,22 @@ def start(self):
...
 
     def run(self):
         try:
             self.start()
-            while self.alive:
-                self.process_limit()
-                self.multi.pipe_ping(self.watchdog_pipe)
-                self.sleep()
-                self.process_work()
+            run_async('Worker %s (%s) work' % (type(self).__name__, self.pid), self._runloop)
```
Comment (antonylesuisse) [inline odoo/service/server.py]: It's the only line needed, so i would just do it manually:
threading.Thread(target=_runloop);t.setDaemon(True);t.start()

File: odoo/service/server.py [hunk 4]
```diff
@@ -822,17 +822,9 @@ def process_limit(self):
...
-        # SIGXCPU (exceeded CPU time) signal handler will raise an exception.
         r = resource.getrusage(resource.RUSAGE_SELF)
         cpu_time = r.ru_utime + r.ru_stime
-        def time_expired(n, stack):
-            _logger.info('Worker (%d) CPU time limit (%s) reached.', self.pid, config['limit_time_cpu'])
-            # We dont suicide in such case
-            raise Exception('CPU time limit exceeded.')
-        signal.signal(signal.SIGXCPU, time_expired)
-        soft, hard = resource.getrlimit(resource.RLIMIT_CPU)
+        _, hard = resource.getrlimit(resource.RLIMIT_CPU)
```
Comment (antonylesuisse) [inline odoo/service/server.py]: leave soft and gain one line of diff

File: odoo/service/server.py [hunk 5]
```diff
@@ -856,17 +849,22 @@ def start(self):
         signal.signal(signal.SIGCHLD, signal.SIG_DFL)
         signal.set_wakeup_fd(self.wakeup_fd_w)
 
+        set_limit_memory_hard()
```
Comment (antonylesuisse) [inline odoo/service/server.py]: Yes it's cleaner to move those one shots here.

File: odoo/service/server.py [hunk 6]
```diff
@@ -1127,3 +1134,22 @@ def restart():
         threading.Thread(target=_reexec).start()
     else:
         os.kill(server.pid, signal.SIGHUP)
+
+def run_async(threadname, func, *args, **kwargs):
```
Comment (antonylesuisse) [inline odoo/service/server.py]: We can remove this.

File: odoo/service/server.py (hunk 1 above)
Comment (xmo-odoo) [reply odoo/service/server.py]: the task specifically mentions preload as an issue (https://www.odoo.com/web#id=1913129&model=project.task&view_type=form)

**Comment by xmo-odoo** [comment]
superseded by #30688

**Pull request closed without merging**